snapshot/snapshotfs/upload_test.go

```diff
@@ -273,3 +282,53 @@ func TestUpload_SubDirectoryReadFailureIgnoreFailures(t *testing.T) {
 		t.Errorf("unexpected directory tree, diff(-got,+want): %v\n", diff)
 	}
 }
+
+// nolint:gocyclo
+func TestUploadWithCheckpointing(t *testing.T) {
+	ctx := testlogging.Context(t)
+	th := newUploadTestHarness(ctx)
+
+	defer th.cleanup()
+
+	u := NewUploader(th.repo)
+
+	policyTree := policy.BuildTree(nil, policy.DefaultPolicy)
+
+	si := snapshot.SourceInfo{
+		UserName: "user",
+		Host:     "host",
+		Path:     "path",
+	}
+
+	count := 0
+
+	// when reading d1 advanced the time, to trigger checkpoint
+	th.sourceDir.Subdir("d1").OnReaddir(func() {
+		count++
+
+		// when reading this directory advance the time to cancel the snapshot
+		// and trigger a checkpoint
+		if count <= 2 {
+			th.ft.Advance(DefaultCheckpointInterval + 1)
+		}
+	})
+
+	if _, err := u.Upload(ctx, th.sourceDir, policyTree, si); err != nil {
+		t.Errorf("Upload error: %v", err)
+	}
+
+	snapshots, err := snapshot.ListSnapshots(ctx, th.repo, si)
+	if err != nil {
+		t.Fatalf("error listing snapshots: %v", err)
+	}
+
+	if got, want := len(snapshots), 2; got != want {
+		t.Fatalf("unexpected number of snapshots: %v, want %v", got, want)
+	}
+
+	for _, sn := range snapshots {
+		if got, want := sn.IncompleteReason, IncompleteReasonCheckpoint; got != want {
```

Comment:
oh, that's because Upload() itself does not write any manifest for completed snapshots. That allows the caller to add things like description, etc. May be worth revisiting, but that's what it is.